modules/rostests/apitests/ntdll/NtNotifyChangeMultipleKeys.c

```diff
@@ -0,0 +1,397 @@
+    if (NT_SUCCESS(Status))
+    {
+        NtDeleteKey(SecondaryKey);   
+        CloseHandle(SecondaryKey);
+    }
+
+    return STATUS_SUCCESS;
+}
+
+NTSTATUS WINAPI NtNotifyChangeMultipleKeys_InitializePerTest(PState state)
+{
+    NTSTATUS Status;
+
+    Status = NtOpenKey(&state->KeyHandle, KEY_SET_VALUE | KEY_NOTIFY | DELETE, &state->ObjectAttributes);
+    CHECK_ERROR(state, Status);
+
+    Status = NtOpenKey(&state->SubKeyHandle, KEY_SET_VALUE | KEY_NOTIFY | DELETE, &state->SubKeyObjectAttributes);
+    CHECK_ERROR(state, Status);
+
+    return STATUS_SUCCESS;
+}
+
+/* Tests */
+
+START_SUBTEST(Synchronous)
+{
+    NTSTATUS Status;
+
+    UNICODE_STRING ValueName;
+    RtlInitUnicodeString(&ValueName, L"TestValue");
+
+    DWORD ValueData1 = 0x12345678;
+    DWORD ValueData2 = 0x87654321;
+
+    INIT_TEST(state, Status);
+    
+    /* Create a thread */
+    state->ThreadHandle = CreateThread(NULL, 0, NtNotifyChangeMultipleKeys_WatchThread, state->KeyHandle, 0, NULL);
+    if (!state->ThreadHandle)
+    {
+        CLEANUP(state);
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    /* Verify the thread is still running */
+    Status = WaitForSingleObject(state->ThreadHandle, 100);
+    TEST_ERROR(state, Status, WAIT_TIMEOUT);
+
+    /* Make change to the registry key */
+    Status = NtSetValueKey(state->KeyHandle, &ValueName, 0, REG_DWORD, &ValueData1, sizeof(ValueData1));
+    CHECK_ERROR(state, Status);
+    /* Give system some time to process the change and notify our watch thread */
+    SleepEx(100, TRUE);
+    Status = NtSetValueKey(state->KeyHandle, &ValueName, 0, REG_DWORD, &ValueData2, sizeof(ValueData2));
+    CHECK_ERROR(state, Status);
+
+    /* Verify that the thread is notified */
+    Status = WaitForSingleObjectEx(state->ThreadHandle, 100, TRUE);
+    TEST_ERROR(state, Status, WAIT_OBJECT_0);
+
+    FINALIZE_TEST(state);
+}
+
+START_SUBTEST(Asynchronous)
+{
+    NTSTATUS Status;
+
+    UNICODE_STRING ValueName;
+    RtlInitUnicodeString(&ValueName, L"AsyncTestValue");
+
+    DWORD ValueData1 = 0x12345678;
+    DWORD ValueData2 = 0x87654321;
+
+    INIT_TEST(state, Status);
+
+    /* Create event */
+    state->EventHandle = CreateEvent(NULL, FALSE, FALSE, NULL);
+    ok(state->EventHandle != NULL, "Failed to create event\n");
+
+    /* Start watching for changes */
+    Status = NtNotifyChangeMultipleKeys(state->KeyHandle, 0, NULL, state->EventHandle, NULL, NULL, &state->IoStatusBlock, REG_NOTIFY_CHANGE_LAST_SET, FALSE, NULL, 0, TRUE);
+    TEST_ERROR(state, Status, STATUS_PENDING);
+
+    /* Check event state */
+    Status = WaitForSingleObject(state->EventHandle, 0);
+    TEST_ERROR(state, Status, WAIT_TIMEOUT);
+
+    /* Make change to the registry key */
+    Status = NtSetValueKey(state->KeyHandle, &ValueName, 0, REG_DWORD, &ValueData1, sizeof(ValueData1));
+    TEST_ERROR(state, Status, STATUS_SUCCESS);
+    SleepEx(100, TRUE);
```

Comment:
I suppose you're waiting for 0.1 s for the CM to set up notification stuff or why that you set a value to the key two times? Explain thoroughly in comments what's happening here.

Reply:
I don't know neither😂 I tried different things on Windows and this was the setup that actually worked.
Without the second call, or without the wait, the change is not notified for whatever reason. Could be a synchronization or a timing issue.

Reply:
Seems like call to Sleep blocks the notification for whatever reason, so instead I used SleepEx with Alertable set to TRUE, hoping that it would work but it didn't neither.

Reply:
IMHO mixing up `NtNotifyChangeMultipleKeys` with `NtSetValueKey` and `WaitForSingleObjectEx` together in one single thread execution is a weird way to be notified of changes in a key. Why don't you just split it into two threads, with thread A setting up the notifications and then it waits on the event that the kernel is going to alert you, and thread B toys around with the key's data?

If you want to properly synchronize the threads together so that you get proper test results, you could synchronize your two threads with a dummy event. So basically thread A sets up the notifications, wait on the notification event, thread B sets up new values to the target key and waits on thread A to receive the notification event and then thread A signals the dummy event to the thread B.

Reply:
I did this for testing the synchronous mode, since it blocks the execution so we won't be able to make changes in registry.
As for asynchronous mode, I think the point of having asynchronous mode is being able to use it in a single-threaded environment, so I think it makes more sense to keep the test for asynchronous mode single threaded.

Reply:
Maybe I can fix this by increasing watcher thread's priority?